### PR TITLE
Keep manual throttle mix at zero throttle when airmode is on

### PR DESCRIPTION
**Pull request by andyp1per**
Should fix https://discuss.ardupilot.org/t/airmode-and-throttle-mixing-is-there-a-bug/86419

### CODE REVIEW

**Comment by IamPete1** [comment]
I think this will prevent, landing detection? So you might find stick disarming is hard.

https://github.com/ArduPilot/ardupilot/blob/c65425e3ac2c9757b3b6eb70a9e8410275c853e3/ArduCopter/land_detector.cpp#L74-L75

**Comment by andyp1per** [comment]
I'll see how it works out in practice

**Comment by andyp1per** [comment]
Actually we need one other small change to allow the landing detector to trigger

File: ArduCopter/land_detector.cpp [hunk 1]
```diff
@@ -72,7 +72,9 @@ void Copter::update_land_detector()
                                     || ((!force_flying || landing) && motors->limit.throttle_lower && pos_control->get_vel_desired_cms().z < 0.0f);
 #else
         // check that the average throttle output is near minimum (less than 12.5% hover throttle)
-        bool motor_at_lower_limit = motors->limit.throttle_lower && attitude_control->is_throttle_mix_min();
+        bool motor_at_lower_limit = motors->limit.throttle_lower
+            // throttle is never at mix min in airmode
+            && (attitude_control->is_throttle_mix_min() || air_mode == AirMode::AIRMODE_ENABLED);
```
Comment (tridge) [inline ArduCopter/land_detector.cpp:77]: needs to only make this change if in a manual throttle mode, otherwise increases risk of land-detect mid-air in modes like LAND

**Review by tridge — DISMISSED** [review]

File: ArduCopter/land_detector.cpp (hunk 1 above)
Comment (andyp1per) [reply ArduCopter/land_detector.cpp]: Fixed

File: ArduCopter/land_detector.cpp [hunk 2]
```diff
@@ -70,9 +70,13 @@ void Copter::update_land_detector()
         const bool landing = flightmode->is_landing();
         bool motor_at_lower_limit = (flightmode->has_manual_throttle() && (motors->get_below_land_min_coll() || heli_flags.coll_stk_low) && fabsf(ahrs.get_roll()) < M_PI/2.0f)
                                     || ((!force_flying || landing) && motors->limit.throttle_lower && pos_control->get_vel_desired_cms().z < 0.0f);
+        bool motor_at_minimum_mix = true;
 #else
         // check that the average throttle output is near minimum (less than 12.5% hover throttle)
-        bool motor_at_lower_limit = motors->limit.throttle_lower && attitude_control->is_throttle_mix_min();
+        bool motor_at_lower_limit = motors->limit.throttle_lower;
+        // throttle is never at mix min in airmode
+        bool motor_at_minimum_mix = attitude_control->is_throttle_mix_min() ||
```
Comment (IamPete1) [inline ArduCopter/land_detector.cpp:78]: Randy is a the naming expert. But the param is called `motor_at_minimum_mix` it returns true in airmode but the comment says it is never at min in airmode, so one would expect it to be false in that case.

File: libraries/AP_HAL_ChibiOS/HAL_ChibiOS_Class.cpp [hunk 3]
```diff
@@ -197,7 +197,7 @@ thread_t* get_main_thread()
 }
 
 static AP_HAL::HAL::Callbacks* g_callbacks;
-
+int start_the_clock = 0;
```
Comment (IamPete1) [inline libraries/AP_HAL_ChibiOS/HAL_ChibiOS_Class.cpp:200]: unrelated?

**Review by IamPete1 — COMMENTED** [review]
I wonder if it would be worth increasing the land detector time threshold (`LAND_DETECTOR_MAYBE_TRIGGER_SEC`) in acro + airmode case. Were now ignoring the throttle mix, currently it would take a little while for it to slew down from mix manual to mix min when you went to low throttle prior to disarming. Removing that check makes the land detector more sensitive. We could bump the threshold time so that it takes approx the same time as it does now when it has to wait for the throttle mix to slew to min first. Of course that slew time does depend on what values you have set...

File: ArduCopter/land_detector.cpp (hunk 2 above)
Comment (andyp1per) [reply ArduCopter/land_detector.cpp:78]: Thanks fixed

File: libraries/AP_HAL_ChibiOS/HAL_ChibiOS_Class.cpp (hunk 3 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/HAL_ChibiOS_Class.cpp]: Thanks fixed

**Comment by andyp1per** [comment]
> currently it would take a little while for it to slew down from mix manual to mix min when you went to low throttle prior to disarming

I can't see that this is true - we go to mix min in update_throttle_mix() as soon as the input throttle hits zero. So I think the landing detector fires in the same time? In my testing I didn't see anything untoward, but maybe I am wrong.

**Comment by IamPete1** [comment]
`is_throttle_mix_min` checks `_throttle_rpy_mix` against `_thr_mix_min` * 1.25.

https://github.com/ArduPilot/ardupilot/blob/3935971e4606cb9c0d02d90816d0eb102ad9e24e/libraries/AC_AttitudeControl/AC_AttitudeControl_Multi.h#L72-L73

`_throttle_rpy_mix` is set in this function where is slewed towards the desired throttle mix as set with the `set_throttle_mix_...` functions.

https://github.com/ArduPilot/ardupilot/blob/3935971e4606cb9c0d02d90816d0eb102ad9e24e/libraries/AC_AttitudeControl/AC_AttitudeControl_Multi.cpp#L315-L327

Taking the slew down example from the comment, 0.9 to 0.1 in 1.6 seconds, the slew down time in that case dwarfs the `LAND_DETECTOR_MAYBE_TRIGGER_SEC` of 0.2 seconds. So you would have a minimum land detection time of (1.6*0.75 + 0.2) 1.4 seconds, with this change that becomes 0.2 seconds. Of course the default mix man and mix min are both 0.1, there is no change. But I think those who use acro might well have a mix man of more then 0.9? So would have had a longer slew time and this change becomes more significant.

Of course as you say in practice this may make no difference, just something to keep a eye out for in testing.

**Comment by andyp1per** [comment]
Isn't LAND_DETECTOR_MAYBE_TRIGGER_SEC  1? But yes I have MIX_MAN at 4...

**Comment by andyp1per** [comment]
@IamPete1 I agree its worth increasing the timeout in air mode, however I don't think we can change LAND_DETECTOR_MAYBE_TRIGGER_SEC as I think this needs to be short in case of multiple attempts. We could increase LAND_DETECTOR_TRIGGER_SEC to 2s, say?

**Comment by andyp1per** [comment]
I've increased the timeout to 3s in airmode

File: ArduCopter/mode.cpp [hunk 4]
```diff
@@ -287,14 +287,16 @@ bool Copter::set_mode(Mode::Number mode, ModeReason reason)
...
 
+    // perform any cleanup required by previous flight mode
+    exit_mode(flightmode, new_flightmode);
+
     if (!new_flightmode->init(ignore_checks)) {
         mode_change_failed(new_flightmode, "initialisation failed");
+        // return to previous flight mode doing proper initialization
+        flightmode->init(true);
         return false;
     }
```
Comment (IamPete1) [inline ArduCopter/mode.cpp:298]: I fear unforeseen consequences with this one. At the least is is quite a big change in behavior to re-enter the current mode. RTL will do climb min again for example. I suspect its the ones that we don't think of that will catch us tho....

Comment (andyp1per) [reply ArduCopter/mode.cpp]: Removed

File: libraries/AC_AttitudeControl/AC_AttitudeControl_Multi.cpp [hunk 5]
```diff
@@ -322,6 +322,19 @@ void AC_AttitudeControl_Multi::update_throttle_rpy_mix()
     } else if (_throttle_rpy_mix > _throttle_rpy_mix_desired) {
         // reduce more slowly (from 0.9 to 0.1 in 1.6 seconds)
         _throttle_rpy_mix -= MIN(0.5f * _dt, _throttle_rpy_mix - _throttle_rpy_mix_desired);
+
+        // if the mix is still higher than that being used, reset immediately
+        float throttle_hover = _motors.get_throttle_hover();
+        float throttle_out = _motors.get_throttle_out();
+        float throttle_in = _motors.get_throttle();
+        float mix_used = 0.0f;
+        if (throttle_out < throttle_hover) {
```
Comment (IamPete1) [inline libraries/AC_AttitudeControl/AC_AttitudeControl_Multi.cpp:331]:
```suggestion
        const float throttle_hover = _motors.get_throttle_hover();
        const float throttle_in = _motors.get_throttle();
        const float throttle_out = max(_motors.get_throttle_out(), throttle_in);
        float mix_used;
        if (throttle_out < throttle_hover) {
```
This in insures that throttle out is always equal or larger to throttle in, and thus throttle out in must always be less than throttle hover, and therefore we never get a divide by zero.

Comment (IamPete1) [reply libraries/AC_AttitudeControl/AC_AttitudeControl_Multi.cpp:331]: A big comment pointing at the equation this is derived from would be good to. And one there so we don't forget to update this if we change it.

**Review by IamPete1 — APPROVED** [review]
Code changes look good. The throttle mix changes will speedup landing detection in all cases, so we should probably do a fair bit of testing. I think the change is correct, if we find landing is detected too soon we should increase the trigger thresholds.

File: ArduCopter/land_detector.cpp [hunk 6]
```diff
@@ -70,9 +71,16 @@ void Copter::update_land_detector()
         const bool landing = flightmode->is_landing();
         bool motor_at_lower_limit = (flightmode->has_manual_throttle() && (motors->get_below_land_min_coll() || heli_flags.coll_stk_low) && fabsf(ahrs.get_roll()) < M_PI/2.0f)
                                     || ((!force_flying || landing) && motors->limit.throttle_lower && pos_control->get_vel_desired_cms().z < 0.0f);
+        bool mix_ready_for_landing = true;
 #else
         // check that the average throttle output is near minimum (less than 12.5% hover throttle)
-        bool motor_at_lower_limit = motors->limit.throttle_lower && attitude_control->is_throttle_mix_min();
+        bool motor_at_lower_limit = motors->limit.throttle_lower;
+        bool mix_ready_for_landing = attitude_control->is_throttle_mix_min();
```
Comment (lthall) [inline ArduCopter/land_detector.cpp:78]: mix_ready_for_landing - This name should describe the state. "throttle_mix_at_min" for example.

File: ArduCopter/land_detector.cpp [hunk 7]
```diff
@@ -70,9 +71,16 @@ void Copter::update_land_detector()
...
+        // throttle is never at mix min in airmode
```
Comment (lthall) [inline ArduCopter/land_detector.cpp:79]: Comment like:
// set throttle_mix_at_min to true because throttle is never at mix min in airmode
// increase land_trigger_sec when using airmode

**Review by lthall — APPROVED** [review]
Some minor feedback about naming and some comments but I am happy with this.

File: ArduCopter/land_detector.cpp (hunk 6 above)
Comment (andyp1per) [reply ArduCopter/land_detector.cpp]: Thanks fixed

File: ArduCopter/land_detector.cpp (hunk 7 above)
Comment (andyp1per) [reply ArduCopter/land_detector.cpp]: Thanks - fixed

**Review by khancyr — COMMENTED** [review]
can't judge about the logic, but no code mistake in those changes

**Review by rmackay9 — APPROVED** [review]
Changes to the landing detector are always scary because of the possibility of false-positives which lead to vehicles suddenly disarming in the air but seems like enough qualified brains have devoted effort to this that it will hopefully be OK.

**Comment by rmackay9** [comment]
I haven't included this in Copter-4.2.3-rc1 because I'm worried that it will cause support issues.  I'd prefer we'd leave this to 4.3 which will begin beta testing in a few weeks.  If you strongly disagree please ping me..
